File: ShellGUI/lib/core/dynamic_library.dart
// ignore_for_file: unused_import, library_prefixes

import '/common.dart';
import 'dart:ffi' as ffi;
import '/core/interface.dart' as Interface;
import '/core/symbol.dart';
import '/core/library.dart';

// ----------------

class DynamicLibrary extends Library {

  late SymbolTable        _symbol;
  late ffi.DynamicLibrary _handle;

  // ----------------

  DynamicLibrary(
    String path,
  ) {
    this._handle = ffi.DynamicLibrary.open(normalizeLibraryPath(path));
    this._symbol = SymbolTable();
    this._symbol.version = this._handle.lookup<ffi.NativeFunction<Interface.version>>(SymbolNameTable.version);
    this._symbol.execute = this._handle.lookup<ffi.NativeFunction<Interface.execute>>(SymbolNameTable.execute);
    this._symbol.prepare = this._handle.lookup<ffi.NativeFunction<Interface.prepare>>(SymbolNameTable.prepare);
  }

  // ----------------

  @override
  ffi.Pointer<Interface.Size>
  version(
  ) {
    return this._symbol.version.asFunction<Interface.version>()();
  }

  @override
  ffi.Pointer<Interface.String>
  execute(
    ffi.Pointer<Interface.Callback>   callback,
    ffi.Pointer<Interface.String>     script,
    ffi.Pointer<Interface.StringList> argument,
  ) {
    return this._symbol.execute.asFunction<Interface.execute>()(callback, script, argument);
  }

  @override
  ffi.Pointer<Interface.String>
  prepare(
  ) {
    return this._symbol.prepare.asFunction<Interface.prepare>()();
  }

}

File: Helper/Core/DynamicLibrary.cs
#pragma warning disable 0,
// ReSharper disable

using Helper;
using ExternalLibrary = Helper.ExternalLibrary;

namespace Helper.Core {

	public unsafe class DynamicLibrary : Library {

		#region structor

		private SymbolTable mSymbol;
		private IntPtr      mHandle;

		// ----------------

		~DynamicLibrary (
		) {
			ExternalLibrary.Kernel32.FreeLibrary(this.mHandle);
		}

		// ----------------

		public DynamicLibrary (
			String path
		) {
			this.mHandle = ExternalLibrary.Kernel32.LoadLibrary($"{path}.");
			this.mSymbol = new SymbolTable() {
				version = (Interface.version)Marshal.GetDelegateForFunctionPointer(ExternalLibrary.Kernel32.GetProcAddress(this.mHandle, SymbolNameTable.version), typeof(Interface.version)),
				execute = (Interface.execute)Marshal.GetDelegateForFunctionPointer(ExternalLibrary.Kernel32.GetProcAddress(this.mHandle, SymbolNameTable.execute), typeof(Interface.execute)),
				prepare = (Interface.prepare)Marshal.GetDelegateForFunctionPointer(ExternalLibrary.Kernel32.GetProcAddress(this.mHandle, SymbolNameTable.prepare), typeof(Interface.prepare)),
			};
		}

		#endregion

		#region interface

		public override Interface.String* Version (
			Interface.Size** number
		) {
			return this.mSymbol.version(number);
		}

		public override Interface.String* Execute (
			Interface.Callback**   callback,
			Interface.String**     script,
			Interface.StringList** argument,
			Interface.String**     result
		) {
			return this.mSymbol.execute(callback, script, argument, result);
		}

		public override Interface.String* Prepare (
		) {
			return this.mSymbol.prepare();
		}

		#endregion

	}

}
